### PR TITLE
Dev/parser

### PR DESCRIPTION
**Pull request by haukurb**

### CODE REVIEW

File: src/greynirseq/nicenlp/utils/label_schema/label_schema.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            ret_mask[cat_vec_idx, schema.group_names.index(group_name)] = 1
+        assert cat_label_idx != dictionary.unk()
+    for cat in schema.label_categories:
+        cat_label_idx = dictionary.index(cat)
+        assert cat_label_idx != dictionary.unk()
+    return ret_mask
+
+
+def make_mapped_group_masks(schema, ldict, device="cpu"):
+    # deprecated
```
Comment (vesteinn) [inline src/greynirseq/nicenlp/utils/label_schema/label_schema.py:67]: Any reason to keep it? Also the one below

File: src/greynirseq/nicenlp/utils/label_schema/label_schema.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    group_name_to_group_attr_vec_idxs = {
+        name: torch.tensor(
+            [dict_.index(item) - offset for item in name_to_labels[name]]
+        )
+        for name in group_names
+    }
+    return group_name_to_group_attr_vec_idxs
+
+
+def make_dict_idx_to_vec_idx(dictionary, cats, device="cpu", fill_value=-100):
```
Comment (vesteinn) [inline src/greynirseq/nicenlp/utils/label_schema/label_schema.py:105]: Also in parser_criterion.py ?

**Review by vesteinn — COMMENTED** [review]
Might be worth it to go for type annotations and mypy at this point

File: src/greynirseq/nicenlp/criterions/pos_criterion.py [hunk 3]
```diff
@@ -9,34 +9,6 @@
...
-
-    no_bos_mask = torch.cat([targets.new_empty(bsz, 1).fill_(pad_idx), targets], 1).ne(
-        pad_idx
-    )
-
-    return no_bos_mask[padding_mask.bitwise_not()]
-
 
 @register_criterion("pos_ice")
 class POSCriterion(FairseqCriterion):
```
Comment (vesteinn) [inline src/greynirseq/nicenlp/criterions/pos_criterion.py:14]: Nothin pos specific in this criterion?

File: src/greynirseq/nicenlp/tasks/pos_task.py [hunk 4]
```diff
@@ -17,84 +17,31 @@
...
-            ret_mask[cat_vec_idx, schema.group_names.index(group_name)] = 1
-        assert cat_label_idx != dictionary.unk()
-    for cat in schema.label_categories:
-        cat_label_idx = dictionary.index(cat)
-        assert cat_label_idx != dictionary.unk()
-    return ret_mask
+logger = logging.getLogger(__name__)
 
 
 @register_task("pos_ice")
```
Comment (vesteinn) [inline src/greynirseq/nicenlp/tasks/pos_task.py:47]: Nothing pos specific about this task?

File: src/greynirseq/nicenlp/criterions/pos_criterion.py [hunk 5]
```diff
@@ -48,11 +20,9 @@ def forward(self, model, sample, reduce=True):
...
         3) logging outputs to display while training
         """
-        assert (
-            hasattr(model, "classification_heads")
-            # and "multi_label_word_classification" in model.classification_heads
-            and "pos_ice" in model.classification_heads
-        ), "model must provide sentence classification head for --criterion=pos_ice"
+        assert hasattr(
+            model, "pos_head"
+        ), "model must provide sentence classification head for pos_ice"
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/criterions/pos_criterion.py:25]: Agreed, "pos_is" or "pos_isl"?

File: src/greynirseq/nicenlp/data/datasets.py [hunk 6]
```diff
@@ -268,29 +270,24 @@ class WordEndMaskDataset(BaseWrapperDataset):
     def __init__(
         self,
         dataset,
+        dictionary,
         is_word_initial,
-        has_bos=True,
-        include_bos=True,
-        include_eos=False,
+        bos_value=1,
+        eos_value=0,
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/data/datasets.py:276]: Was refactored out and should've been removed.

File: src/greynirseq/nicenlp/data/datasets.py [hunk 7]
```diff
@@ -268,29 +270,24 @@ class WordEndMaskDataset(BaseWrapperDataset):
...
-        # exclude specials for now
-        starts = [self.is_word_initial.get(int(v), 1) for v in item[offset:-1]]
-
-        mask = torch.zeros_like(item)
-        mask[:offset] = 1 if self.include_bos else 0  # if exists
-        mask[-1] = 1 if self.include_eos else 0
-        mask[offset:-1] = torch.tensor(starts[1:] + [1]).type_as(mask)
-
+        mask = torch.tensor([self.is_word_initial.get(int(v), 1) for v in item])
+        mask[self.start_offset] = 1  # temporary hack due to incorrect preprocessing (missing prepend_space)
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/data/datasets.py:288]: GPT2's tokenizer prepends space to the input (if there is no leading space). This is to make the first token be word-initial. The parameter for prepending leading space for the HuggingFace tokenizer used was false (should've been true).

File: src/greynirseq/nicenlp/models/pos_model.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+from greynirseq.nicenlp.models.multilabel_word_classification import (
+    MultiLabelTokenClassificationHead
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+@register_model("icebert_pos")
+class IceBERTPOS(RobertaModel):
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/models/pos_model.py:31]: I think so.

File: src/greynirseq/nicenlp/models/pos_model.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+        word_mask = sample["net_input"]["word_mask"]
+        nwords = sample["net_input"]["word_mask"].sum(-1)
+
+        (cat_logits, attr_logits), _extra = self.model(tokens, word_mask=word_mask)
+
+        return self.task.logits_to_labels(cat_logits, attr_logits, word_mask)
+
+    def predict(self, sentences, device="cuda"):
+        device = torch.device(device)
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/models/pos_model.py:162]: It's used when debugging and manual queries from the REPL.

File: src/greynirseq/nicenlp/models/simple_parser.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        if args.freeze_embeddings:
+            freeze_module_params(sentence_encoder.embed_tokens)
+            freeze_module_params(sentence_encoder.segment_embeddings)
+            freeze_module_params(sentence_encoder.embed_positions)
+            freeze_module_params(sentence_encoder.emb_layer_norm)
+
+        for layer in range(args.n_trans_layers_to_freeze):
+            freeze_module_params(sentence_encoder.layers[layer])
+
+        self.task = task
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/models/simple_parser.py:114]: Only twice repeated doesn't justify a subclass IMO.

File: src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py [hunk 11]
```diff
@@ -267,21 +234,14 @@ def build_model(self, args):
...
-        #     num_classes_binary = self.args.num_classes_binary
-
-        model.register_classification_head(
-            "multi_span_classification",
-            # num_cats=len(self.label_schema.label_categories),
-            num_cats=None,
-            num_labels=self.num_labels,
-        )
+        raise NotImplementedError
+        # TODO: This is incorrect
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py:238]: Hasn't been implemented yet.

File: src/greynirseq/nicenlp/tasks/parser_task.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+        self.dictionary = data_dictionary
+
+        self.nterm_dictionary = nterm_dict
+        self.nterm_schema = nterm_schema
+
+        if not hasattr(args, "max_positions"):
+            self._max_positions = (args.max_source_positions, args.max_target_positions)
+        else:
+            self._max_positions = args.max_positions
+        args.tokens_per_sample = self._max_positions
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/parser_task.py:93]: That is intentional by fairseq (it has resolve_max_positions).

File: src/greynirseq/nicenlp/tasks/parser_task.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    @classmethod
+    def load_dictionary(cls, args, filename, add_mask=True):
+        """Load the dictionary from the filename
+
+        Args:
+            filename (str): the filename
+        """
+        dictionary = Dictionary.load(filename)
+        if add_mask:
+            dictionary.add_symbol("<mask>")
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/parser_task.py:140]: It could, but this is in line with fairseqs (it's implicitly defined in multiple places). Where do you think it should be defined?

File: src/greynirseq/nicenlp/tasks/parser_task.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+    @classmethod
+    def get_word_beginnings(cls, args, dictionary):
+        bpe = encoders.build_bpe(args)
+        if bpe is not None:
+
+            def is_beginning_of_word(i):
+                if i < dictionary.nspecial:
+                    return True
+                tok = dictionary[i]
+                if tok.startswith("madeupword"):
```
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/parser_task.py:152]: This is part of how the Dictionary class is defined, this is padding (so that its length is divisible by 8).

File: src/greynirseq/nicenlp/utils/label_schema/label_schema.py (hunk 1 above)
Comment (haukurb) [reply src/greynirseq/nicenlp/utils/label_schema/label_schema.py:67]: Should've been removed.

File: src/greynirseq/nicenlp/criterions/pos_criterion.py (hunk 3 above)
Comment (haukurb) [reply src/greynirseq/nicenlp/criterions/pos_criterion.py:14]: Agreed, changing name to MultiLabelTokenClassificationCriterion, along with its filename.

File: src/greynirseq/nicenlp/tasks/pos_task.py (hunk 4 above)
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/pos_task.py:47]: Agreed, renaming and changing filename as well.

File: src/greynirseq/nicenlp/models/pos_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/models/pos_model.py:31]: What about the name? The model is specific to classifying tokens in a sequence but aside from that it is task-agnostic (NER, POS, etc).

File: src/greynirseq/nicenlp/data/datasets.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/data/datasets.py:276]: Disregard the previous comment. 

This is the value in the mask (if it occurs in the input string). The token classification tasks use bos_value=0 and the parser task uses bos_value=1 (the span that includes only the first word needs the bos vector).

File: src/greynirseq/nicenlp/models/pos_model.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/models/pos_model.py:158]: The variable `device` is not used in the method

File: src/greynirseq/nicenlp/tasks/parser_task.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/tasks/parser_task.py:147]: Was thinking dictionary.mask or something along those lines, similar to bos, eos

File: src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py:230]: Maybe clarify with a comment and remove the broken stuff? Or is the plan to fix it before the PR goes through?

[the remark this answers is not in the document]
Comment (haukurb) [reply src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py:230]: The plan wasn't to fix it for this PR, it was left as is to be potentially used for grammatical error detection.

[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/tasks/multilabel_multispan_classification.py:230]: Leave a comment explaining that?

File: src/greynirseq/nicenlp/tasks/parser_task.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/tasks/parser_task.py:159]: Divisible by 8? Length of madeupword is 10. Was thinking if this literal could be read from somewhere

File: src/greynirseq/nicenlp/models/simple_parser.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vesteinn) [reply src/greynirseq/nicenlp/models/simple_parser.py:106]: Two things to change, easy to miss one of them if this needs updating, e.g. if embed_tokens gets renamed embedded_tokens

File: src/greynirseq/nicenlp/models/simple_parser.py [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+@register_model("icebert_simple_parser")
+class SimpleParserModel(RobertaModel):
+    """Simple chart parser model based on Roberta. Simple in the sense that it does not
+       add additional self-attentive layers or feature-engineering from Kitaev and Klein 2018."""
+
+    def __init__(self, args, encoder, task):
+        super().__init__(args, encoder)
+
+        def freeze_module_params(m):
+            if m is not None:
```
Comment (vesteinn) [inline src/greynirseq/nicenlp/models/simple_parser.py:91]: Do we really want to continue if m is None?

Comment (haukurb) [reply src/greynirseq/nicenlp/models/simple_parser.py:91]: There is nothing else in the stub after the if statement.
